baseplate/diagnostics/tracing/publisher.py

```diff
@@ -0,0 +1,191 @@
+RETRY_LIMIT_DEFAULT = 10
+
+ZIPKIN_API_VERSION_DEFAULT = "v1"
+ZIPKIN_SCHEME_DEFAULT = "http"
+
+
+class RetryLimitExceeded(Exception):
+    pass
+
+
+class TraceBatch(V1Batch):
+    def __init__(self, max_size):
+        super(TraceBatch, self).__init__(max_size=max_size)
+
+
+class ZipkinPublisher(object):
+    """Zipkin trace publisher."""
+    def __init__(self, endpoint,
+            metrics_client,
+            post_timeout=POST_TIMEOUT_DEFAULT,
+            zipkin_api_version=ZIPKIN_API_VERSION_DEFAULT,
+            zipkin_scheme=ZIPKIN_SCHEME_DEFAULT,
+            retry_delay=RETRY_DELAY_DEFAULT,
+            retry_limit=RETRY_LIMIT_DEFAULT,
+            num_conns=5):
+
+        adapter = requests.adapters.HTTPAdapter(pool_connections=num_conns,
+                                                pool_maxsize=num_conns)
+        self.session = requests.Session()
+        self.session.mount("{}://".format(zipkin_scheme), adapter)
+        self.endpoint = "{}://{}/api/{}/spans".format(zipkin_scheme, endpoint, zipkin_api_version)
+        self.metrics = metrics_client
+        self.post_timeout = post_timeout
+        self.retry_delay = retry_delay
+        self.retry_limit = retry_limit
+
+    def publish(self, payload):
+        """Publish spans to Zipkin API.
+
+        :param baseplate.events.publisher.SerializedBatch payload: Count and
+            payload to publish.
+        """
+        if not payload.count:
+            return
+
+        logger.info("Sending batch of %s traces", payload.count)
+        headers = {
+            "User-Agent": "baseplate-trace-publisher/1.0",
+            "Content-Type": "application/json",
+        }
+        tries = 0
+        while True:
+            try:
+                with self.metrics.timer("post"):
+                    response = self.session.post(
+                        self.endpoint,
+                        data=payload.bytes,
+                        headers=headers,
+                        timeout=self.post_timeout,
+                        stream=False
+                    )
+                response.raise_for_status()
+            except requests.HTTPError as exc:
+                self.metrics.counter("error.http").increment()
+                response = getattr(exc, "response", None)
+                if response is not None:
+                    logger.exception("HTTP Request failed. Error: %s", response.text)
+                    # If client error, crash
+                    if response.status_code < 500:
+                        raise
+                else:
+                    logger.exception("HTTP Request failed. Response not available")
+            except IOError:
+                self.metrics.counter("error.io").increment()
+                logger.exception("HTTP Request failed")
+            else:
+                self.metrics.counter("sent").increment(payload.count)
+                return
+
+            tries += 1
+            if tries > self.retry_limit:
```

Comment:
Maybe this could use https://baseplate.readthedocs.io/en/stable/baseplate/retry.html#baseplate.retry.RetryPolicy ?